### PR TITLE
oci-image: init scripts to build and upload image

### PR DESCRIPTION
**Pull request by ilian**
###### Motivation for this change
Add image configuration for Oracle Cloud Infrastructure and scripts to
build and upload the image as a [Custom Image](https://docs.oracle.com/en-us/iaas/Content/Compute/Tasks/importingcustomimagelinux.htm#).

###### Things done

- [x] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [x] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: nixos/maintainers/scripts/oci/create-image.sh
Comment (SuperSandro2000) [inline nixos/maintainers/scripts/oci/create-image.sh:11]:
```suggestion
   -j$NIX_BUILD_CORES
```
Can this be used here?

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (SuperSandro2000) [inline nixos/maintainers/scripts/oci/upload-image.sh:1]: I would prefer using bash explicitly because /bin/sh is bash on nixpkgs but dash/ash/etc on other systems.

File: nixos/maintainers/scripts/oci/create-image.sh
Comment (ilian) [reply nixos/maintainers/scripts/oci/create-image.sh:11]: It seems that NIX_BUILD_CORES is only available inside certain nix environments such as inside a nix-shell but not outside of those environments. I think it would be better to set the maximum number of jobs to auto, which should achieve the same result without relying on the environment.

Comment (SuperSandro2000) [reply nixos/maintainers/scripts/oci/create-image.sh]: The reason we are normally not doing this is that if you have 32 cores you might not want to use them all for one build.

Comment (ilian) [reply nixos/maintainers/scripts/oci/create-image.sh]: Alright, should I try to read NIX_BUILD_CORES if it exists and fall back to a constant (e.g. 8) instead? Or should I omit the option altogether to fall back to the option defined in nix.conf?

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (grahamc) [inline nixos/maintainers/scripts/oci/upload-image.sh:1]: For what it is worth, after working with the AWS uploader for some time, I have needed to reimplement it with Terraform to keep track of the details. I wonder if using Terraform for this instead of a script would work too?

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (grahamc) [inline nixos/maintainers/scripts/oci/upload-image.sh:69]: This is not very much metadata. Can we add more, or am I missing it? For example:

* exact version of the channel used to build it
* any relevant filesystem information
* who built it and when
* perhaps a version number for the metadata itself, so we could add more types of images another time and not break automated integrations

File: nixos/modules/virtualisation/oci-image.nix
Comment (grahamc) [inline nixos/modules/virtualisation/oci-image.nix:27]: Should this service be enabled only on the very first profile? I wonder if this should be part of the -user config.

**Comment by samueldr** [comment]
Hi!

Expect to see some eyes on your PR, since that announcement about their A1 instances.

On my end I worked out how to adapt this for those A1 instances

 - https://github.com/samueldr/nixpkgs/commits/feature/ocl-image-aarch64

(Look carefully at the commits, the tip adds a "bad" commit to work around an issue with RK3399... That one is not needed.)

I don't know if there's anything else that should be done here though. I'm *definitely not* an OCI expert. I basically started learning about their cloud offering Wednesday.

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (ilian) [reply nixos/maintainers/scripts/oci/upload-image.sh:1]: I think Terraform would work for this too. The oci terraform provider seems to have the functionality needed to import a custom image: https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_image

Would you mind sharing what details needed to be tracked in your use-case for AWS? I can't currently write/test Terraform code as my free trial has expired which unfortunately prevents me from importing custom images.

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (ilian) [reply nixos/maintainers/scripts/oci/upload-image.sh:69]: Yes, I agree. I think this can be achieved similar to how EC2 images are built by creating a JSON manifest when buliding the image and passing this data to OCI: https://github.com/NixOS/nixpkgs/blob/6933d068c5d2fcff398e802f7c4e271bbdab6705/nixos/maintainers/scripts/ec2/amazon-image.nix#L86-L92

I think we should add some basic metadata such as the channel version without any metadata version number to the display name to avoid clutter, and include all verbose metadata to the Custom Image itself with [Resource Tags](https://docs.oracle.com/en-us/iaas/Content/General/Concepts/resourcetags.htm) which can be provided in the JSON format with the `oci` cli tool.

File: nixos/modules/virtualisation/oci-image.nix
Comment (ilian) [reply nixos/modules/virtualisation/oci-image.nix:27]: The idea here is that the service is executed at least once before the initial ssh connection. This service does indeed need to only run once, since the ssh keys can't be modified in OCI after creating the instance AFAIK.
It should indeed be enabled only for the first profile. The user is allowed to change the system configuration to not include the `fetch-ssh-keys` service, since all ssh keys should be present in `/root/.ssh/authorized_keys` by then, hence this service is not included in `oci-config-user.nix`.

**Comment by ilian** [comment]
> Hi!
> 
> Expect to see some activity on your PR, since that announcement about their A1 instances.

That's great news! Thank you for making me aware of their new ARM instances.

> On my end I worked out how to adapt this for those A1 instances
> 
>     * https://github.com/samueldr/nixpkgs/commits/feature/ocl-image-aarch64
Thanks! I'll add your commits to this PR.

> You will have to ensure that the _image capabilities_ only includes UEFI as the firmware type.

Can this manual configuration step be avoided after modifying the `--launch-mode PARAVIRTUALIZED` to `--launch-mode NATIVE`? IIRC, this should enable UEFI by default. The reason I used PARAVIRTUALIZED for x86 is that it was mentioned as a requirement in https://docs.oracle.com/en-us/iaas/Content/Compute/Tasks/importingcustomimagelinux.htm 
I think the default image of Oracle Linux uses UEFI-based images so we might want to try building UEFI-based images for both x86 and AArch64 to keep things simple.

Unfortunately, my Free Trial has expired so I am not able to test out any new additions to the build script since the `custom-image-count` limit has been reverted to 0 after my trial ended.
I will try to install NixOS on a system with the Oracle Linux image in-place for users that have a free account like mine using https://github.com/elitak/nixos-infect or something similar.

**Comment by samueldr** [comment]
Just a reminder:

> I'm definitely not an OCI expert. I basically started learning about their cloud offering less than 48 hours ago.

> > You will have to ensure that the _image capabilities_ only includes UEFI as the firmware type.
> 
> Can this manual configuration step be avoided after modifying the `--launch-mode PARAVIRTUALIZED` to `--launch-mode NATIVE`? IIRC, this should enable UEFI by default on x86, but I'm not sure what the behavior is for ARM-based instances. The reason I used PARAVIRTUALIZED for x86 is that BIOS is a requirement for custom images as mentioned in https://docs.oracle.com/en-us/iaas/Content/Compute/Tasks/importingcustomimagelinux.htm
> It seems that the current Oracle Linux image for AArch64 (Oracle-Linux-7.9-aarch64-2021.05.17-0) has its launch mode set to PARAVIRTUALIZED instead of NATIVE and its firmware set to UEFI though. If we want to automate setting UEFI as the preferred firmware, we can do so using Terraform (see [launch_options](https://registry.terraform.io/providers/hashicorp/oci/latest/docs/resources/core_image#launch_options)) or a python script that uses oci-sdk. I can't seem to find an option of `oci-cli` that allows me to customize the `LaunchOptions` of an image, even though the manpage of `oci compute image import --help` references `LaunchOptions`. I have opened an issue upstream: [oracle/oci-cli#416](https://github.com/oracle/oci-cli/issues/416)

As far as I understand OCI, the "Launch Mode" is distinct from "Firmware". And as the documentation indeed states, "BIOS" seemingly *needs* to be set for `x86_64-linux` (though I can't confirm).

Given the following link:

 - https://docs.oracle.com/en-us/iaas/Content/Compute/Tasks/configuringimagecapabilities.htm

I wonder if it's as simple as adding `--firmware UEFI_64`. But that's a total guess. 

> I think the default image of Oracle Linux on x86 uses UEFI-based images so we might want to try building UEFI-based images for both x86 and AArch64 to keep things simple.

It would be nice to see which shapes *actually run on UEFI*. I am betting that the documentation stating it **needs** to be legacy boot is plainly wrong.

> Unfortunately, my Free Trial has expired so I am not able to test out any new additions to the build script since the `custom-image-count` limit has been reverted to 0 after my trial ended.
> I will try to install NixOS on a system with the Oracle Linux image in-place for users that have a free account like mine using https://github.com/elitak/nixos-infect or something similar.

Oh what??? The custom images are not part of the free offer?? I wasn't finding any information about the pricing for those, like at all... This sucks :|

So I guess going with "infect" like systems will be better in the end. Unless the NixOS foundation somehow gets in touch with Oracle to get some kind of useful deal.

**Comment by samueldr** [comment]
Reading the (admittedly thick to navigate) [pricing list](https://www.oracle.com/cloud/price-list.html#compute-vm), coupled with the [free tied page](https://www.oracle.com/ca-en/cloud/free/#always-free), I'm led to believe there is up to 10GB available free for custom images.

Expect an edit to this message, I'm chatting with a live agent about this.

**Comment by Djelibeybi** [comment]
Hey folks, I just replied to @samueldr (I assume, based on the avatar) over on the Oracle Cloud forum. I'm a member of the Oracle Cloud Developer Adoption team as well as one of Oracle's GitHub admins. I'm more than happy to be your official tour guide to OCI and make sure you have the right tools and contacts to get NixOS published to our Marketplace.

The first thing I'll mention is that upgrading your free trial account to pay-as-you-go is (ironically) one of the requirements to activate the Always Free resources, which will then increase your custom image count back up to 25. You will not be charged for any resources that qualify for Always Free, and those that do will have an "Always Free" tag visible in the Console.

**Comment by grahamc** [comment]
Hi Avi, thanks for reaching out! Is it possible and useful for the NixOS
foundation to publish official images for Oracle Cloud? Ideally
individual users would not need to maintain their own base images out of
the gate. 

On May 29, 2021, GitHub ***@***.***> wrote:
> Hey folks, I just replied to @samueldr <https://github.com/samueldr>
> (I assume, based on the avatar) over on the Oracle Cloud forum. I'm a
> member of the Oracle Cloud Developer Adoption team as well as one of
> Oracle's GitHub admins. I'm more than happy to be your official tour
> guide to OCI and make sure you have the right tools and contacts to
> get NixOS published to our Marketplace.
>
> The first thing I'll mention is that upgrading your free trial account
> to pay-as-you-go is (ironically) one of the requirements to activate
> the Always Free resources, which will then increase your custom image
> count back up to 25. You will not be charged for any resources that
> qualify for Always Free, and those that do will have an "Always Free"
> tag visible in the Console.
>

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/NixOS/nixpkgs/pull/119856#issuecomment-850923986>,
or unsubscribe <https://github.com/notifications/unsubscribe-
auth/AAASXLEEQZQ2CO6MVFRVWRTTQGHW3ANCNFSM43FSFLTA>.

**Comment by Djelibeybi** [comment]
It's both useful and possible to publish NixOS images to the [Oracle Cloud Marketplace](https://cloudmarketplace.oracle.com/marketplace/en_US/homeLinkPage). This will make it available to all Oracle Cloud users via the Create Instance dialog on the Console, as well as via the API or SDK.

File: nixos/modules/virtualisation/oci-image.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/oci-image.nix:46]:
```suggestion
      serviceConfig = {
        Type = "oneshot";
        RemainAfterExit = true;
        StandardError = "journal+console";
        StandardOutput = "journal+console";
      };
```

File: nixos/modules/virtualisation/oci-image.nix
Comment (SuperSandro2000) [inline nixos/modules/virtualisation/oci-image.nix:42]:
```suggestion
      script = ''
        mkdir -m 0700 -p /root/.ssh
        if [ -f /root/.ssh/authorized_keys ]; then
          echo "Authorized keys have already been downloaded"
        else
          echo "Downloading authorized keys from Instance Metadata Service v2"
          curl -s -S -L \
            -H "Authorization: Bearer Oracle" \
            -o /root/.ssh/authorized_keys \
            http://169.254.169.254/opc/v2/instance/metadata/ssh_authorized_keys
          chmod 600 /root/.ssh/authorized_keys
        fi
      '';
```

the chmod should probably retry a couple of times.

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (SuperSandro2000) [inline nixos/maintainers/scripts/oci/upload-image.sh:92]:
```suggestion
if [[ -n $par_id ]]; then
```

File: nixos/modules/virtualisation/oci-image.nix
Comment (ilian) [reply nixos/modules/virtualisation/oci-image.nix]: Why should the chmod operation be retried?

**Comment by mweinelt** [comment]
Using kexec via https://github.com/cleverca22/nix-tests/blob/master/kexec/session.md worked for me to get into a live system. Unfortunately it rebooted into the EFI shell after a while. I terminated the previous instance and created a new one and disabled the cloud agent in the instance configuration. I assume it works like a watchdog, so remember to disable it if you try that.

**Comment by zhaofengli** [comment]
For aarch64, can we use systemd-boot as the bootloader? On boot, GRUB gives me the following error:

```
file `/boot/grub/arm64-efi/efi_uga.​mod' not found.
```

It boots, but it's not possible to select the system generation through the serial console. systemd-boot seems to work fine from my testing.

**Comment by samueldr** [comment]
@zhaofengli in theory yes. In practice last I tried our systemd-boot tooling failed on AArch64.

Though what should be done, instead or in addition, is fix the GRUB issue. It goes a bit more deeply than that specific warning/error. Serial vs. graphical GRUB are not working correctly with the current configuration, and I haven't looked at the actual underlying issue.

**Comment by zhaofengli** [comment]
> @zhaofengli in theory yes. In practice last I tried our systemd-boot tooling failed on AArch64.

I tried and it worked fine, though I ran `bootctl install` instead of `NIXOS_INSTALL_BOOTLOADER=1`. Can't reboot right now but I'd guess the normal way should also work.

> Though what should be done, instead or in addition, is fix the GRUB issue. It goes a bit more deeply than that specific warning/error. Serial vs. graphical GRUB are not working correctly with the current configuration, and I haven't looked at the actual underlying issue.

I think it's fixable by simply removing `insmod efi_uga` from `nixos/modules/system/boot/loader/grub/install-grub.pl` for aarch64 and other platforms on which it isn't built. UGA was removed in UEFI in favor of GOP, but it's still in use on EFI 1.x platforms (e.g., some Apple devices).

**Comment by samueldr** [comment]
Though I'm 99% sure this is not the *reason* you don't see the graphical menu with GRUB. So yes, it probably should be removed, but I expect it will change nothing.

File: nixos/modules/virtualisation/oci-common.nix
Comment (colemickens) [inline nixos/modules/virtualisation/oci-common.nix:7]: what does OCI use from this?

Comment (ilian) [reply nixos/modules/virtualisation/oci-common.nix:7]: The profile contains the necessary kernel modules for several QEMU/VirtIO devices. Running `lspci` on my instance running on OCI confirms that there are several emulated VirtIO devices present:
```
00:00.0 Host bridge: Red Hat, Inc. QEMU PCIe Host bridge
00:01.0 Display controller: Red Hat, Inc. Virtio GPU (rev 01)
00:02.0 USB controller: Red Hat, Inc. QEMU XHCI Host Controller (rev 01)
00:03.0 Ethernet controller: Red Hat, Inc. Virtio network device
00:10.0 PCI bridge: Red Hat, Inc. QEMU PCIe Root port
...
18:00.0 SCSI storage controller: Red Hat, Inc. Virtio SCSI (rev 01)
```

**Comment by K900** [comment]
Maybe we could rewrite the scripts to Python instead of bash? Python has more API access and better controls than the CLI.

File: nixos/maintainers/scripts/oci/upload-image.sh
Comment (TilCreator) [inline nixos/maintainers/scripts/oci/upload-image.sh:20]: This call is missing the required argument of `create-image.sh`

**Comment by PedroHLC** [comment]
In case someone wonders, this PR still works quite well (14 months after the latest contribution).

- I've rebased the branch with latest `release-22.05`, without conflicts;
- Configured my `~/.oci` using OpenSSL commands provided in Oracle's documentation;
- Ran `./create-image.sh aarch64-linux`, it created the file without problems;
- Ran `./update-image.sh`, it automatically uploaded the file, created a temporary bucket, transformed the file in a custom image, as expected;
- Then it asked me to change the shape compatibility in a link, I've also used the chance to change the boot capability to only-UEFI (tried with BIOS in an earlier attempt, and it didn't boot);
- Meanwhile, the script took it time but deleted the file and bucket without destroying the custom image;
- Created an instance from it: new virtual network and new subnet, disabled all OS management plugins, and added my keys by uploading their `.pub` files;
- Once I had the public IP, I logged through SSH using the root user and one of the private keys.

Are you guys still interested in merging it?

**Comment by dit7ya** [comment]
Will love to see this merged!

**Comment by puffnfresh** [comment]
I've been using the `oci-common.nix` module for a few months now. Working really well!

**Comment by chenlijun99** [comment]
I also would love to see this merged.

Hi @puffnfresh. By any chance, are you using Oracle's ARM VPS? If yes, didn't you cross-compile the image for "aarch64-linux" on a "x86_64-linux" machine? I'm asking because I tried to copy the modules of this PR and then to cross-build an image, but I get the error:

```
error: a 'aarch64-linux' with features {} is required to build '/nix/store/8ifcgasp770hf6ffc98kz3wrn23bcspc-nixos-enter.drv', but I am a 'x86_64-linux' with features {benchmark, big-parallel, kvm, nixos-test}
```

If you did something similar, I would really appreciate some suggestions. Thanks!

**Comment by puffnfresh** [comment]
@chenlijun99 I probably did, can't remember. I do have access to a few arm64 systems so I might have just compiled the image from there.

It looks like you're not actually cross-compiling. You probably want to set system to `x86_64-linux` and then set `nixpkgs.crossSystem`.

**Comment by thiagokokada** [comment]
Rebased this PR and did a bunch of fixes:
- Use `set -euo pipefail` in the scripts, so if they fail for any reason the script will stop instead of continuing
- Remove deprecated options
- Always use EFI by default, because this is how most images are configured in OCI nowadays. There is still the option to use BIOS if you want
- Set `networking.useNetworkd = true`, because otherwise the `fetch-ssh-key.service` would fail for me, resulting in a VM without the proper SSH keys

I think this should fix most of the basic issues in this PR. Tested and it is working fine in my personal account.

There is a few outstanding other issues, but I think we could merge this PR as-is and work on then in the future, since this PR already has some value as-is (e.g.: people wanting to play with NixOS in Oracle Cloud, specially because of their generous Always Free tier).

**Comment by K900** [comment]
Code looks fine, unfortunately I can't test it because of trade embargo.

**Comment by thiagokokada** [comment]
I did 2 tests:
- Booting the generated `x86_64-linux` image in the `VM.Standard.E2.1.Micro` instance
- Booting the generated `aarch64-linux` image in the `VM.Standard.A1.Flex` instance

Both took a while to be reachable through SSH (a few minutes VS a few seconds in the official images from OCI), but they work well enough to give most people a start point.

While most people are probably more interested in the `aarch64-linux` images, I have had more success using the [Kexec tarballs from nixos-anywhere](https://github.com/nix-community/nixos-images#kexec-tarballs) (contrary on what the README says, you can use those images like a Live environment without nixos-anywhere). I also tested nixos-anywhere itself and it works fine, and probably my recommended way if you want to do so.

However, for the `VM.Standard.E2.1.Micro` the Kexec is simply too big to work, so this image still makes sense for those cases. Keep in mind that the 1GB of RAM restriction really hurts most NixOS setups, and you will need probably need to setup some swap space to make anything build in this environment.

File: nixos/modules/virtualisation/oci-common.nix
Comment (PedroHLC) [inline nixos/modules/virtualisation/oci-common.nix:23]: Does it work having all the three? I know it works for aarch64 because it's what I used/use, but I didn't see this before.

**Review by PedroHLC — APPROVED** [review]
Code still LGTM

File: nixos/modules/virtualisation/oci-common.nix
Comment (thiagokokada) [reply nixos/modules/virtualisation/oci-common.nix:23]: Yes: https://www.kernel.org/doc/html/latest/admin-guide/serial-console.html

> It is possible to specify multiple devices for console output. You can define a new kernel command line option to select which device(s) to use for console output.

**Comment by thiagokokada** [comment]
Going to merge this. Future fixes can go after merging it.

**Comment by cjemorton** [comment]
Does anyone have a simple step by step tutorial on how to use this?

**Pull request merged**
